pkg/cloud/services/compute/instance.go

```diff
@@ -188,23 +188,20 @@ func (s *Service) createInstance(eventObject runtime.Object, clusterName string,
 			return nil, fmt.Errorf("no network was found or provided. Please check your machine configuration and try again")
 		}
 
+		iTrunk := instance.Trunk
+		iTags := instance.Tags
+
+		// if port is created implicitly and trunk is enabled at instance level, then inherit it.
+		if len(*instance.Networks) == 1 && iTrunk && network.PortOpts == nil {
```

Comment:
this condition seems to be problematic, should not it be:
```suggestion
		if len(*instance.Networks) == 1 && iTrunk == nil && network.PortOpts == nil {
```
or 

```suggestion
		if len(*instance.Networks) == 1 || (iTrunk && network.PortOpts) == nil {
```
?

Reply:
+1 for the latter. 
@macaptain Is it possible to add a default `PortOpts` for the port that is created implicitly ? 
Basically, we have the following.
```
func (s *Service) createInstance(eventObject,clusterName,instance)
case 1: provided no port
- Since no port is specified one port is created implicitly.
- The trunk value for this port is taken from openstackMachine.spec.trunk (nil at the moment)

networkID: 655c0dde-17ea-446e-841a-7b19f7b85cda
  port 0: basic-1-control-plane-4k2v4-0

case 2: provided one user defined port
networkID: ca5e7f81-8f9c-46d0-a739-c274be7927ad:
  port 0: basic-1-control-plane-cgx7g-0

case 3: provided three user defined ports
networkID: 3994ffe5-b3c4-4ea3-9b01-ba21edad3aa2
  port 0: basic-1-control-plane-qvwdb-0
  port 1: basic-1-control-plane-qvwdb-1
  port 2: basic-1-control-plane-qvwdb-2
```

Also, relevant is the call to `replaceAllAttributesTags`. When there were no multiple ports, the tags from OpenStackMachine.Spec.Tags was added on the tags of the trunk (which was also only one). However, with addition of multiple ports, there is an ambiguity as to where to put these tags. I see two options, 

1. Put the tag on port[0]
2. Add tags field to the PortOpts

Reply:
Yes, you can set `PortOpts` for the default port in the `constructNetworks` function here:

https://github.com/Nordix/cluster-api-provider-openstack/blob/a2acda37a6aaca788e52e2b641b8ac8a0b6a9f4c/pkg/cloud/services/compute/instance.go#L170-L176

Regarding `replaceAllAttributesTags`, it looks as though the instance tags are set on the default/instance trunk by this function. This isn't documented anywhere from what I can see, I am not sure I would expect this behaviour by default. So I'd leave it alone and not do this for any port level trunk. We can add tags to the PortOpts if needed (could be a follow-up task.)